src/opencmw_worker/src/NilmDataWorker.hpp

```diff
@@ -0,0 +1,208 @@
+using opencmw::NoUnit;
+
+// context for InferenceTool
+struct NilmAcquisitionContext {
+    int64_t                 lastRefTrigger = 0;
+    opencmw::MIME::MimeType contentType    = opencmw::MIME::JSON;
+};
+
+ENABLE_REFLECTION_FOR(NilmAcquisitionContext, lastRefTrigger, contentType)
+
+// data for InferenceTool
+struct AcquisitionNilm {
+    std::vector<int64_t> refTriggerStamp;
+    std::vector<float>   apparentPowerSpectrumStridedValues; // fft(S)
+    std::vector<float>   realPower;                          // P
+    std::vector<float>   reactivePower;                      // Q
+    std::vector<float>   apparentPower;                      // S
+    std::vector<float>   phi;
+};
+
+ENABLE_REFLECTION_FOR(AcquisitionNilm, refTriggerStamp, apparentPowerSpectrumStridedValues, realPower, reactivePower, apparentPower, phi)
+
+struct PQSPhiData {
+    int64_t timestamp     = 0;
+    float   realPower     = 0.0f; // P
+    float   reactivePower = 0.0f; // Q
+    float   apparentPower = 0.0f; // S
+    float   phi           = 0.0f;
+};
+
+struct UISData {
+    int64_t            timestamp = 0;
+    std::vector<float> voltageSpectrum;
+    std::vector<float> currentSpectrum;
+    std::vector<float> apparentPowerSpectrum;
+};
+
+using namespace opencmw::disruptor;
+using namespace opencmw::majordomo;
+template<units::basic_fixed_string serviceName, typename... Meta>
+class NilmDataWorker : public Worker<serviceName, NilmAcquisitionContext, Empty, AcquisitionNilm, Meta...> {
+    PQSPhiData          _timeData;
+    std::mutex          _timeDataMutex;
+    static const size_t RING_BUFFER_SIZE = 256;
+    struct RingBufferData {
+        UISData    freqData;
+        PQSPhiData timeData;
+    };
+    using ringbuffer_t = std::shared_ptr<RingBuffer<RingBufferData, RING_BUFFER_SIZE, BusySpinWaitStrategy, SingleThreadedStrategy>>;
+    using sequence_t   = std::shared_ptr<Sequence>;
+    ringbuffer_t _nilmDataBuffer;
+    sequence_t   _nilmDataBufferTail;
+
+public:
+    using super_t = Worker<serviceName, NilmAcquisitionContext, Empty, AcquisitionNilm, Meta...>;
+
+    template<typename BrokerType>
+    explicit NilmDataWorker(const BrokerType &broker)
+        : super_t(broker, {}), _nilmDataBuffer(newRingBuffer<RingBufferData, RING_BUFFER_SIZE, BusySpinWaitStrategy, ProducerType::Single>()), _nilmDataBufferTail(std::make_shared<Sequence>()) {
+        _nilmDataBuffer->addGatingSequences({ _nilmDataBufferTail });
+
+        // register callback only for "P@100Hz,Q@100Hz,S@100Hz,phi@100Hz"
+        std::scoped_lock lock_time(gr::pulsed_power::globalTimeSinksRegistryMutex);
+        for (auto sink : gr::pulsed_power::globalTimeSinksRegistry) {
+            const auto               signal_names = sink->get_signal_names();
+            const auto               signal_units = sink->get_signal_units();
+            const auto               sample_rate  = sink->get_sample_rate();
+
+            std::vector<std::string> pqsohi_str{ "P", "Q", "S", "phi" };
```

Comment:
Here is a tiny typo :)

Reply:
Typo fixed [here](https://github.com/fair-acc/pulsed-power-ml/pull/129/commits/a8a08523e9991e73e95cab2d8e7cf52914c69eac) :)